pkg/liqo-controller-manager/storageprovisioner/remoteprovisioner.go

```diff
@@ -13,3 +13,145 @@
+		func(pvc *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+			return mutateRemoteRealPVC(virtualPvc, pvc, remoteRealStorageClass, remoteNamespace)
+		}); err != nil {
+		return nil, controller.ProvisioningInBackground, err
+	}
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: options.PVName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			StorageClassName: options.StorageClass.Name,
+			AccessModes:      options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceStorage: options.PVC.Spec.Resources.Requests[v1.ResourceStorage],
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				HostPath: &v1.HostPathVolumeSource{
+					Path: "/tmp/liqo-placeholder",
+				},
+			},
+			NodeAffinity: &v1.VolumeNodeAffinity{
+				Required: &v1.NodeSelector{
+					NodeSelectorTerms: []v1.NodeSelectorTerm{
+						{
+							MatchExpressions: []v1.NodeSelectorRequirement{
+								{
+									Key:      "kubernetes.io/hostname",
+									Operator: v1.NodeSelectorOpIn,
+									Values:   []string{options.SelectedNode.Name},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if options.StorageClass.ReclaimPolicy != nil {
+		pv.Spec.PersistentVolumeReclaimPolicy = *options.StorageClass.ReclaimPolicy
+	}
+
+	return pv, controller.ProvisioningFinished, nil
+}
+
+// mutateRemoteRealPVC enforces the remote realPVC fields merging its old values (if any)
+// with the ones coming from the virtualPVC.
+// i.e. the PVC spec is the copy of the virtual one, but the storage class is the one set in the
+// storage provisioner or the one previously set in the PVC (since it is a read-only field). The
+// real volumeName is preserved too.
+func mutateRemoteRealPVC(virtualPvc, realPvc *v1.PersistentVolumeClaim,
+	storageClass, namespace string) (*v1.PersistentVolumeClaim, error) {
+	if realPvc == nil {
+		realPvc = &v1.PersistentVolumeClaim{
+			TypeMeta: virtualPvc.TypeMeta,
+		}
+	}
+
+	realPvc.ObjectMeta.Name = virtualPvc.Name
+	realPvc.ObjectMeta.Namespace = namespace
+
+	if realPvc.ObjectMeta.Labels == nil {
+		realPvc.ObjectMeta.Labels = map[string]string{}
+	}
+
+	for k, v := range forge.ReflectionLabels() {
+		realPvc.ObjectMeta.Labels[k] = v
+	}
+
+	if realPvc.ObjectMeta.Annotations == nil {
+		realPvc.ObjectMeta.Annotations = map[string]string{}
+	}
+
+	realPvName := realPvc.Spec.VolumeName
+	realPvc.Spec = *virtualPvc.Spec.DeepCopy()
+	realPvc.Spec.VolumeName = realPvName
+
+	if storageClass == "" {
+		realPvc.Spec.StorageClassName = nil
+	} else {
+		realPvc.Spec.StorageClassName = &storageClass
+	}
+
+	return realPvc, nil
+}
+
+func createOrUpdate(ctx context.Context,
+	lister corev1listers.PersistentVolumeClaimNamespaceLister,
+	client corev1clients.PersistentVolumeClaimInterface,
+	objName string, f func(pcv *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error)) (*v1.PersistentVolumeClaim, error) {
```

Comment:
Could use server-side-apply to simplify this logic.

Reply:
I added a commit keeping both the solutions, the servers-side-apply seems to me to be too verbose (don't look at the names of the functions, I have to change it).

I don't know which version to keep

Reply:
I personally don't find it more verbose than the other alternative (possibly even less, considering you would remove the create or update part). Could also shortened a bit, removing some new lines... Still, that is mostly a matter of personal taste (I personally prefer the SSA alternative).